<a href="https://colab.research.google.com/github/eviemcmahan/ds2002-fa26/blob/main/notebooks/03-pandas-cleaning/2026-09-18%20%E2%80%94%20Pandas%20Challenge%20%E2%80%94%20Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DS2002 · Pandas Challenge

**Lab — 2026-09-18 · Fall 2026**  

---

## Lab 04 — Pandas Challenge

Four hundred generated orders. Each question builds toward a demand report you could hand a vendor.

The data is seeded, so everyone's numbers should match. That is deliberate: if your total revenue differs from your neighbor's, one of you has a bug, and the assertions at the end will tell you which.

Every answer needs the number **and** a sentence saying what it means. A cell that prints `4218.5` with no interpretation is half an answer.

In [67]:
import pandas as pd, numpy as np
rng = np.random.default_rng(4)
n = 400
df = pd.DataFrame({
    'vendor_id': rng.choice(['V-01','V-05','V-10','V-18'], n),
    'category': rng.choice(['Food','Merch','RainGear','Drink'], n, p=[.5,.2,.1,.2]),
    'qty': rng.integers(1, 4, n),
    'price': rng.choice([4.5, 6.0, 7.5, 12.0, 24.0], n),
})
df.head()

,vendor_id,category,qty,price
0,V-10,Drink,2,24.0
1,V-18,RainGear,1,12.0
2,V-18,Drink,3,4.5
3,V-10,Food,2,12.0
4,V-18,Drink,3,7.5


### Q1 — Add `revenue`, then report total revenue and total units.

*Expected: 400 rows, and revenue should land between $8,000 and $9,000.*

In [68]:
df['revenue'] = df['qty'] * df['price'] # Add revenue by multiplying qty by price
print(f"total revenue: ${df['revenue'].sum():.2f}")
print("total units:", df['revenue'].count())


total revenue: $8520.00
total units: 400


The total revenue and units matched what I expected. Since I'm using a dataframe, I can just do calculations on each variable.

### Q2 — Revenue by category, highest to lowest.

Include the share of total as a percentage in the same table.

In [69]:
df_by_category = df.groupby('category').sum() # Group by category
df_by_category['revenue_share'] = (df_by_category['revenue'] / df_by_category['revenue'].sum() * 100).round(2) # Calculate share of total as % of revenue
df_by_category = df_by_category.sort_values('revenue', ascending=False) # Sort by revenue

print(df_by_category)

                                                  vendor_id  qty   price  \
category                                                                   
Food      V-10V-05V-10V-10V-01V-10V-01V-10V-18V-18V-10V-...  362  2187.0   
Merch     V-18V-01V-05V-18V-18V-05V-05V-05V-05V-01V-10V-...  158   888.0   
Drink     V-10V-18V-18V-18V-10V-05V-18V-18V-01V-18V-01V-...  178   814.5   
RainGear  V-18V-05V-05V-05V-18V-01V-01V-10V-18V-01V-05V-...   85   513.0   

          revenue  revenue_share  
category                          
Food       4293.0          50.39  
Merch      1771.5          20.79  
Drink      1554.0          18.24  
RainGear    901.5          10.58  


Food has the highest revenue of $4,293, followed by merch at $1,771.50, drink at $1,554, and rain gear at $901.50. After grouping by category, I could then calculate the percentage of revenue share which I did by dividing the each category's revenue sum by the total amount of revenue generated.

### Q3 — Which vendor has the highest *average* order revenue?

Report the average alongside the order count for each vendor. A high average on twelve orders is a different claim from a high average on two hundred.

In [70]:
# TODO

### Q4 — What share of revenue comes from Merch?

Print it as a percentage rounded to one decimal.

In [71]:
# TODO

### Q5 — Join in the vendor names.

The frame only has `vendor_id`. Merge the lookup below so your report is readable.

**Requirements:** left join, `validate='many_to_one'`, and prove the row count and revenue total did not change. One vendor id in the orders is not in this lookup — find it, and decide what to do about it.

In [72]:
vendor_names = pd.DataFrame({
    'vendor_id': ['V-01', 'V-05', 'V-10'],
    'vendor_name': ['Hoos Burgers', 'Rotunda Tacos', 'Cav Merch North'],
})

# TODO: merge, validate, and report the unmatched vendor

**The unmatched vendor, and what I did about it:** _..._

### Q6 — A pivot table: vendors down the side, categories across the top, revenue in the cells.

Add row and column totals so it reads as a report rather than a grid of numbers.

In [73]:
# TODO

### Q7 — Validate your work

**TODO:** uncomment and make these pass. Assign your results to the named variables as you go.

In [74]:
# assert len(df) == 400
# assert 8000 < df['revenue'].sum() < 9000, df['revenue'].sum()
# assert abs(by_category['revenue'].sum() - df['revenue'].sum()) < 0.01
# assert len(joined) == len(df), 'the vendor merge changed the row count'
print('checks passed.')

checks passed.


### Write-up

**a)** What would you tell these vendors to do differently next game? One paragraph, with at least two numbers from your report in it.

**b)** Which of your seven answers is the least trustworthy, and why? Point at a specific weakness — a small group size, an unmatched vendor, a category that is really two things.

_your answer here_